### DISCUSSION TITLE
Viewport and the ! flag

### DISCUSSION THREAD

**Discussion by speedgeil** [ImageMagick/ImageMagick#5258] [Help]
I have thousands of documents from the local archive that all have a thick red border around it. These are all scans and the border must be preserved. Unfortunatel the photo contents within the border is bleached out and i want to normalize the content only without the border affecting the normalization process.
As far as i understood from the docs the ! flag will treat a cropped geometry area like a viewport and the resturned result would bring me the border back again after only using the cropped area for processing. But it doesn't. So i guess i misunderstood that.
Thing is, i want the border to be affected by the normalization too, but i don't want Imagemagick "to see" the border while calculating, because it will return false colors then (it tries to neutralize the red of the border).
Anyone knows anything?

**Comment by snibgo** [comment]
What version of IM, on what platform?

Please show the command you are using.

One possibility is to crop out the area you want to change, do the changes you want, and composite just that area back on the original input.

**Reply by speedgeil** [reply]
I'm currently not at my working place. So i'm trying from my mind.
I'm using something like 

magick input.tif -crop 10x10+10+10\! output.tif

I'm on the latest version 7.x on Windows.

I did it with recomposing before but i always had some edge where the difference between both images shows.

**Reply by speedgeil** [reply]
> What version of IM, on what platform?
> 
> Please show the command you are using.
> 
> One possibility is to crop out the area you want to change, do the changes you want, and composite just that area back on the original input.

Could it be possible to run the normalize command on a cropped image and somehow 'save' the calculation result to a hald clut that is then applied to the uncropped image?

**Comment by fmw42** [comment]
That is not easily done, since -normalize has no way to list the gray level values associated with clip values of 3% and 1% that it uses.  In principle, one could accumulate values from each side of the histogram until you reach 3% on the high side and 1% on the low side and get the associated gray levels at those points. Then one could use -level to apply those values to any image you wanted.

**Comment by snibgo** [comment]
You can do your own "-contrast-stretch" by making a histogram of the small area, and applying that "-level" to the entire image. See https://github.com/ImageMagick/ImageMagick/discussions/2383

Or you can do a "-contrast-stretch", and by comparing the input to the output, calculate what "-level" or "-level-colors" does the same task. See https://github.com/ImageMagick/ImageMagick/discussions/4025

**Comment by speedgeil** [comment]
Thank you both for your explanations. What about working with layers? Could we apply normalization to the upper layer and remove it and the changes apply to the bottom layer like in Photoshop?

**Reply by fmw42** [reply]
I do not understand the question. Please explain in more detail?  What do you mean by remove it etc? I think the same issue applies to layers as to images. You will have to do the analysis?

**Perhaps the developers will consider adding -defines to print the graylevel break points corresponding to the specified percent values in -normalize and -contrast-stretch.**

**Reply by speedgeil** [reply]
In Photoshop, when working with pixel and adjustment layers, you can apply an automatic normalization to the pixel layer underneath with the help of a levels layer. 
Assuming we got two pixel layers below that levels layer and we apply the normalization to the first layer below the levels layer, the normalization will also work on the bottom layer after removing the pixel layer inbetween.

**Reply by fmw42** [reply]
We do not have an easy way to get the level values from the -normalize, which uses cumulative percents. If we did, we could apply -level to any layer you wanted.
